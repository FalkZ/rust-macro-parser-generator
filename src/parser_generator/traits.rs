use std::fmt::Debug;

use super::{result::{ParserError, ParserResult}, sourcemap::{Token}};



pub trait OrMessage<V> {
    fn or_message(self, str: &str) -> ParserResult<V>;
}

impl<A> OrMessage<A> for Option<A> {
    fn or_message(self, message: &str) -> ParserResult<A> {
        match self {
            Some(v) => Ok(v),
            None => ParserError::error(message),
        }
    }
}

pub trait OrErr<V> {
    fn or_err(self) -> ParserResult<V>;
}

impl<A, E: Into<anyhow::Error>> OrErr<A> for Result<A, E> {
    fn or_err(self) -> ParserResult<A> {
        match self {
            Ok(v) => Ok(v),
            Err(e) => {
                let er : anyhow::Error = e.into();
                
                ParserError::error( &er.to_string())
            },
        }
    }
}



pub trait OrErrString<V> {
    fn or_err(self) -> ParserResult<V>;
}

 impl<A, E: Into<String>> OrErrString<A> for Result<A, E> {
    fn or_err(self) -> ParserResult<A> {
        match self {
            Ok(v) => Ok(v),
            Err(e) => ParserError::error(&e.into()),
        }
    }
}




pub trait RawToken: Clone {
    fn raw_token(&self) -> Token;
}

impl <T: Into<Token> + Clone> RawToken for T {
    fn raw_token(&self) -> Token {
       self.clone().into()
    }
}


