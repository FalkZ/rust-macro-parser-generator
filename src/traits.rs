use crate::result::{ParserResult, ParserError};

pub trait T<V> {
    fn or_message(self, str: &str) -> ParserResult<V>;
}

impl<A> T<A> for Option<A> {
    fn or_message(self, message: &str) -> ParserResult<A> {
        match self {
            Some(v) => Ok(v),
            None => ParserError::error(message),
        }
    }
}

pub trait T2<V> {
    fn or_err(self) -> ParserResult<V>;
}

impl<A, E: Into<anyhow::Error>> T2<A> for Result<A, E> {
    fn or_err(self) -> ParserResult<A> {
        match self {
            Ok(v) => Ok(v),
            Err(e) => {
                let er : anyhow::Error = e.into();
                
                ParserError::error( &er.to_string())
            },
        }
    }
}


pub trait T3<V> {
    fn or_err(self) -> ParserResult<V>;
}

 impl<A, E: Into<String>> T3<A> for Result<A, E> {
    fn or_err(self) -> ParserResult<A> {
        match self {
            Ok(v) => Ok(v),
            Err(e) => ParserError::error(&e.into()),
        }
    }
}
