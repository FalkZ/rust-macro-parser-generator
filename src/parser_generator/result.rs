use std::error::Error;
use std::fmt;

pub type ParserResult<T> = Result<T, ParserError>;

#[derive(Debug, Clone)]
pub enum ParserError {
    Mismatch,
    Unreachable,
    UnreachableAt(String),
    Err(TextError),
}

impl fmt::Display for ParserError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", &self)
    }
}

impl Error for ParserError {
    fn description(&self) -> &str {
        "Error while parsing"
    }

    fn cause(&self) -> Option<&dyn Error> {
        match &self {
            ParserError::Err(_) => Some(self),
            _ => None,
        }
    }
}


#[derive(Debug, Clone)]
pub struct TextError {
    message: String
}


impl Error for TextError {
    fn description(&self) -> &str {
       &self.message
    }

    fn cause(&self) -> Option<&dyn Error> {
        None 
    }
}

impl fmt::Display for TextError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", &self)
    }
}


impl ParserError {
    pub fn error<T>(message: &str) -> ParserResult<T> {
        let err = TextError{ message: message.to_string() };
        Err(ParserError::Err(err))
    }
}


