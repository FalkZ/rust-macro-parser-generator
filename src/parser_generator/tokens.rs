use std::{
    cell::RefCell,
    fmt::{Debug, Display},
    rc::Rc,
};

use super::position::{GetPosition, Position};

#[derive(Debug, Clone)]

pub struct Pin<'a, T> {
    pinned_index: usize,
    tokens: &'a Tokens<T>,
}

impl<'a, T> Pin<'a, T> {
    pub fn get_pinned(&'a self) -> &'a Tokens<T> {
        let mut r = self.tokens.index.borrow_mut();

        *r = self.pinned_index;

        self.tokens
    }
}

#[derive(Default, Clone, Debug)]
pub struct Token {
    pub raw: String,
    pub position: Position,
}

impl Display for Token {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.raw)
    }
}

impl Token {
    pub fn new(raw: String, position: Position) -> Self {
        Self { raw, position }
    }
    pub fn as_str<'a>(&'a self) -> &'a str {
        &self.raw
    }
}

#[derive(Debug, Clone)]
pub struct Tokens<T> {
    index: Rc<RefCell<usize>>,
    tokens: Vec<T>,
}

impl<T: GetPosition> Tokens<T> {
    pub fn new(tokens: Vec<T>) -> Self {
        Self {
            index: Rc::new(RefCell::new(0)),
            tokens,
        }
    }

    pub fn pin(&self) -> Pin<T> {
        Pin {
            pinned_index: self.index(),
            tokens: self,
        }
    }

    fn index(&self) -> usize {
        let reference = self.index.borrow();
        *reference
    }

    pub fn next(&self) -> Option<&T> {
        let next = self.tokens.get(self.index()).clone();

        let mut reference = self.index.borrow_mut();
        *reference += 1;

        return next;
    }

    pub fn peek(&self) -> Option<&T> {
        let next = self.tokens.get(self.index()).clone();

        return next;
    }

    pub fn position(&self) -> Option<Position> {
        let t = &self.peek()?;

        Some(t.position().clone())
    }
}

impl<T: GetPosition + Display> Display for Tokens<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s: String = self
            .tokens
            .iter()
            .enumerate()
            .map(|(i, t)| {
                if i == self.index() {
                    format!("|> {} <|", t)
                } else {
                    format!("{}", t)
                }
            })
            .collect::<Vec<String>>()
            .join(", ");

        write!(f, "{}", s)
    }
}

pub trait RawToken: Clone {
    fn raw_token(&self) -> Token;
    fn as_str(&self) -> String {
        let t = self.raw_token();

        t.raw
    }
}

impl<T: Into<Token> + Clone> RawToken for T {
    fn raw_token(&self) -> Token {
        self.clone().into()
    }
}
