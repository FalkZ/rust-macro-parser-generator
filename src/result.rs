use std::error::Error;
use std::fmt;

pub type ParserResult<T> = Result<T, ParserError>;

#[derive(Debug)]
pub enum ParserError {
    Mismatch,
    Err(anyhow::Error),
}

impl fmt::Display for ParserError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", &self)
    }
}

impl Error for ParserError {
    fn description(&self) -> &str {
        "Error while parsing"
    }

    fn cause(&self) -> Option<&dyn Error> {
        match &self {
            ParserError::Err(e) => Some(self),
            _ => None,
        }
    }
}

impl ParserError {
    pub fn error<T>(message: &str) -> ParserResult<T> {
        let err = anyhow::anyhow!("{}", message);
        Err(ParserError::Err(err))
    }
}
