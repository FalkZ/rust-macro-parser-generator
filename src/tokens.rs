use std::{cell::RefCell, fmt::Display, rc::Rc};
use itertools::Itertools;

use crate::sourcemap::Pos;

#[derive(Debug, Clone)]

pub struct Pin<'a, T> {
    pinned_index: usize,
    tokens: &'a Tokens<T>,
}

impl<'a, T> Pin<'a, T> {
    pub fn get_pinned(&'a self) -> &'a Tokens<T> {
        let mut r = self.tokens.index.borrow_mut();

        *r = self.pinned_index;

        self.tokens
    }
}

#[derive(Debug, Clone)]
pub struct Tokens<T> {
    index: Rc<RefCell<usize>>,
    tokens: Vec<T>,
}

impl<T: Pos> Tokens<T> {
    pub fn new(tokens: Vec<T>) -> Self {
        Self {
            index: Rc::new(RefCell::new(0)),
            tokens,
        }
    }

    pub fn pin(&self) -> Pin<T> {
        Pin {
            pinned_index: self.index(),
            tokens: self,
        }
    }

    fn index(&self) -> usize {
        let reference = self.index.borrow();
        *reference
    }

    pub fn next(&self) -> Option<&T> {
        let next = self.tokens.get(self.index()).clone();

        let mut reference = self.index.borrow_mut();
        *reference += 1;

        return next;
    }

    pub fn peek(&self) -> Option<&T> {
        let next = self.tokens.get(self.index()).clone();

        return next;
    }

    pub fn position(&self)-> Option<crate::sourcemap::Position> {
        let t = &self.peek()?;
        
        Some(t.position().clone())
    }
}

impl<T: std::fmt::Debug + Pos> Display for Tokens<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s: String = self.tokens.iter().enumerate().map(|(i, t)| {
            if i == self.index() {
                format!("|> {:?} <|", t)
            } else {
                format!("{:?}", t)
            }
        }).intersperse(", ".to_string()).collect();
        
        write!(f, "{}", s)
    }
}