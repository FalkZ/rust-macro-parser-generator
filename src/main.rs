mod lexer;
mod matchers;
mod parser;
mod result;
mod tokens;
mod traits;
//mod example_parser;

use result::{ParserError, ParserResult};
use traits::*;

use crate::tokens::Tokens;

Lexer!(
    {
        {'0'..='9' =>} => NUMBER(i64),
        {'A'..='Z' | 'a'..='z' =>} => IDENT(String),
        {'"' | '\'' => '"' | '\''} => TEXTLITERAL(String),
        {'.'} => DOT,
        {'+'} => PLUS,
        {'-'} => MINUS,
        {'='} => EQUAL,
        {','} => COMMA,
        {'/'} => DIVISION,
        {'*'} => MULTIPLICATION,
        {'('} => BRACKETOPEN,
        {')'} => BRACKETCLOSE,
        {'{'} => CBRACKETOPEN,
        {'}'} => CBRACKETCLOSE
    }

    { ' ' | '\n' } => _

    {
        IDENT => { {"^"} => POWER }
        //NUMBER => { { 123 } => T123(i64) }
    }
);

// op = PLUS | MINUS | POWER | DIV
// _term3 = term | NUMBER(i64)
// term = NUMBER(i64) op _term3

Parser!(
    operator = {PLUS | MINUS | DIVISION | IDENT(String) },
    value = { NUMBER(i64) | TEXTLITERAL(String) | IDENT(String)},
    expression = (#value => value, #operator => operator, #expressions => ex),
    expressions = {#expression | #value},
    argument =  (IDENT(String) => arg,  COMMA, #arguments => rest),
    arguments = {#argument | IDENT(String)},
    function = ( IDENT(String) => name, BRACKETOPEN, #arguments => arguments,  BRACKETCLOSE, EQUAL, #expressions => body),
    assignment = ( IDENT(String) => name, EQUAL, #expressions => body),
    statement = {#function | #assignment}
);

struct V;

#[derive(Debug)]
enum ASTVariants {
    Arguments(Vec<String>),
    Function(Function),
    None,
}

#[derive(Debug)]
struct Function {
    name: String,
    args: Vec<String>,
}

impl V {
    fn arguments(&self, args: &arguments) -> ASTVariants {
        let mut vec: Vec<String> = vec![];

        let mut a = args;

        loop {
            match a {
                arguments::argument(ar) => {
                    vec.push(ar.arg.0.clone());
                    a = &ar.rest
                }
                arguments::IDENT(str) => {
                    vec.push(str.to_owned());
                    break;
                }
            }
        }

        ASTVariants::Arguments(vec)
    }

    fn function(&self, f: &function) -> ASTVariants {
        let name = f.name.0.to_owned();
    
        match self.arguments(&f.arguments) {
            ASTVariants::Arguments(args) => ASTVariants::Function(Function {
                name,
                args,
            }),
            _ => ASTVariants::None,
        }
    }

    fn statement(&self, statement: &statement) -> ASTVariants {
        match statement {
            statement::function(fun) => self.function(fun),
            statement::assignment(_) => todo!(),
        }
    }
}

fn run() -> ParserResult<Box<statement>> {
    let a = "test(a, b) = a + b";

    // let a = "test = a";

    let t = Parser::new(a)?;

    let t = Parser::statement(&t.tokens)?;

    let v = V {};

    println!("{:?}", v.statement(&t));
    Ok(t)
}

fn main() {
    println!("{:?}", run());
}
