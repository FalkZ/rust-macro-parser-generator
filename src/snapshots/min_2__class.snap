---
source: src/main.rs
expression: "&r.tokens"
---
Tokens {
    index: RefCell {
        value: 87,
    },
    tokens: [
        IMPORT(
            Token {
                raw: "import",
                position: Position {
                    column: 0,
                    line: 1,
                },
            },
        ),
        COL(
            Token {
                raw: ":",
                position: Position {
                    column: 6,
                    line: 1,
                },
            },
        ),
        RAWIDENT(
            Token {
                raw: "`@std/util`",
                position: Position {
                    column: 8,
                    line: 1,
                },
            },
        ),
        CBRACKETOPEN(
            Token {
                raw: "{",
                position: Position {
                    column: 19,
                    line: 1,
                },
            },
        ),
        IDENT(
            Token {
                raw: "pipe",
                position: Position {
                    column: 21,
                    line: 1,
                },
            },
        ),
        CBRACKETCLOSE(
            Token {
                raw: "}",
                position: Position {
                    column: 26,
                    line: 1,
                },
            },
        ),
        COMMA(
            Token {
                raw: ",",
                position: Position {
                    column: 27,
                    line: 1,
                },
            },
        ),
        RAWIDENT(
            Token {
                raw: "`@std/math`",
                position: Position {
                    column: 4,
                    line: 2,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 15,
                    line: 2,
                },
            },
        ),
        IDENT(
            Token {
                raw: "PI",
                position: Position {
                    column: 0,
                    line: 3,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 3,
                    line: 3,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 5,
                    line: 3,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 7,
                    line: 3,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 9,
                    line: 3,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 11,
                    line: 3,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "0",
                position: Position {
                    column: 13,
                    line: 3,
                },
            },
        ),
        DOT(
            Token {
                raw: ".",
                position: Position {
                    column: 14,
                    line: 3,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "012",
                position: Position {
                    column: 15,
                    line: 3,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 18,
                    line: 3,
                },
            },
        ),
        PUB(
            Token {
                raw: "pub",
                position: Position {
                    column: 0,
                    line: 5,
                },
            },
        ),
        MUT(
            Token {
                raw: "mut",
                position: Position {
                    column: 4,
                    line: 5,
                },
            },
        ),
        IDENT(
            Token {
                raw: "Pi",
                position: Position {
                    column: 8,
                    line: 5,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 11,
                    line: 5,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "123",
                position: Position {
                    column: 13,
                    line: 5,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 16,
                    line: 5,
                },
            },
        ),
        IDENT(
            Token {
                raw: "fn",
                position: Position {
                    column: 0,
                    line: 7,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 2,
                    line: 7,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 3,
                    line: 7,
                },
            },
        ),
        COMMA(
            Token {
                raw: ",",
                position: Position {
                    column: 4,
                    line: 7,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 6,
                    line: 7,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 7,
                    line: 7,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 9,
                    line: 7,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 11,
                    line: 7,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 13,
                    line: 7,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 15,
                    line: 7,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 16,
                    line: 7,
                },
            },
        ),
        IDENT(
            Token {
                raw: "one",
                position: Position {
                    column: 0,
                    line: 9,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 3,
                    line: 9,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 4,
                    line: 9,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 5,
                    line: 9,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 7,
                    line: 9,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 9,
                    line: 9,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 11,
                    line: 9,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 13,
                    line: 9,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 15,
                    line: 9,
                },
            },
        ),
        IDENT(
            Token {
                raw: "c",
                position: Position {
                    column: 17,
                    line: 9,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 18,
                    line: 9,
                },
            },
        ),
        PUB(
            Token {
                raw: "pub",
                position: Position {
                    column: 0,
                    line: 11,
                },
            },
        ),
        IDENT(
            Token {
                raw: "zero",
                position: Position {
                    column: 4,
                    line: 11,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 8,
                    line: 11,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 9,
                    line: 11,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 11,
                    line: 11,
                },
            },
        ),
        TYPESCRIPT(
            Token {
                raw: "#(()=>{throw new Error(\"hello 2\")})()#",
                position: Position {
                    column: 13,
                    line: 11,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 51,
                    line: 11,
                },
            },
        ),
        IDENT(
            Token {
                raw: "multiline",
                position: Position {
                    column: 0,
                    line: 13,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 9,
                    line: 13,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 10,
                    line: 13,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 12,
                    line: 13,
                },
            },
        ),
        IDENT(
            Token {
                raw: "console",
                position: Position {
                    column: 14,
                    line: 13,
                },
            },
        ),
        DOT(
            Token {
                raw: ".",
                position: Position {
                    column: 21,
                    line: 13,
                },
            },
        ),
        IDENT(
            Token {
                raw: "log",
                position: Position {
                    column: 22,
                    line: 13,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 25,
                    line: 13,
                },
            },
        ),
        TEXTLITERAL(
            Token {
                raw: "\"test space\"",
                position: Position {
                    column: 26,
                    line: 13,
                },
            },
        ),
        COMMA(
            Token {
                raw: ",",
                position: Position {
                    column: 38,
                    line: 13,
                },
            },
        ),
        TEXTLITERAL(
            Token {
                raw: "\"multi\n\n    line\"",
                position: Position {
                    column: 40,
                    line: 13,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 9,
                    line: 15,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 10,
                    line: 15,
                },
            },
        ),
        IDENT(
            Token {
                raw: "nesting",
                position: Position {
                    column: 0,
                    line: 17,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 7,
                    line: 17,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 8,
                    line: 17,
                },
            },
        ),
        EQUAL(
            Token {
                raw: "=",
                position: Position {
                    column: 10,
                    line: 17,
                },
            },
        ),
        IDENT(
            Token {
                raw: "one",
                position: Position {
                    column: 12,
                    line: 17,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 15,
                    line: 17,
                },
            },
        ),
        IDENT(
            Token {
                raw: "fn",
                position: Position {
                    column: 16,
                    line: 17,
                },
            },
        ),
        BRACKETOPEN(
            Token {
                raw: "(",
                position: Position {
                    column: 18,
                    line: 17,
                },
            },
        ),
        IDENT(
            Token {
                raw: "a",
                position: Position {
                    column: 19,
                    line: 17,
                },
            },
        ),
        PLUS(
            Token {
                raw: "+",
                position: Position {
                    column: 21,
                    line: 17,
                },
            },
        ),
        IDENT(
            Token {
                raw: "b",
                position: Position {
                    column: 23,
                    line: 17,
                },
            },
        ),
        MINUS(
            Token {
                raw: "-",
                position: Position {
                    column: 25,
                    line: 17,
                },
            },
        ),
        IDENT(
            Token {
                raw: "c",
                position: Position {
                    column: 27,
                    line: 17,
                },
            },
        ),
        COMMA(
            Token {
                raw: ",",
                position: Position {
                    column: 28,
                    line: 17,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "123",
                position: Position {
                    column: 30,
                    line: 17,
                },
            },
        ),
        DOT(
            Token {
                raw: ".",
                position: Position {
                    column: 33,
                    line: 17,
                },
            },
        ),
        NUMBER(
            Token {
                raw: "2",
                position: Position {
                    column: 34,
                    line: 17,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 35,
                    line: 17,
                },
            },
        ),
        BRACKETCLOSE(
            Token {
                raw: ")",
                position: Position {
                    column: 36,
                    line: 17,
                },
            },
        ),
        SEMI(
            Token {
                raw: ";",
                position: Position {
                    column: 37,
                    line: 17,
                },
            },
        ),
    ],
}
